### PR TITLE
fixed some race conditions

### PR DESCRIPTION
**Pull request by sebastianbaginski**

### CODE REVIEW

File: source/concurrent/ThreadPool.ooc [hunk 1]
```diff
@@ -50,7 +50,10 @@ _Task: abstract class {
 			}
 			timer free()
 		}
-		this _state == _PromiseState Finished
+		this _mutex lock()
+		status = this _state == _PromiseState Finished
+		this _mutex unlock()
+		status
```
Comment (sebastianbaginski) [inline source/concurrent/ThreadPool.ooc]: This one was reported when running helgrind on `ThreadPoolTest`

File: source/concurrent/Promise.ooc [hunk 2]
```diff
@@ -71,18 +71,19 @@ ConditionPromise: class extends Promise {
...
 	}
 	wait: override func (time: TimeSpan) -> Bool { Debug error("Timed wait is not supported for ConditionPromise"); false }
 	wait: override func ~forever -> Bool {
 		this _mutex lock()
 		if (!this _completed)
 			this _condition wait(this _mutex)
+		status := this _completed
 		this _mutex unlock()
-		this _completed
+		status
```
Comment (sebastianbaginski) [inline source/concurrent/Promise.ooc]: This one was not reported, but if we're protecting `_completed` with the mutex, let's always do that.

File: source/concurrent/Promise.ooc [hunk 3]
```diff
@@ -71,18 +71,19 @@ ConditionPromise: class extends Promise {
...
 	}
 	signal: func {
-		this _mutex with(||
-			this _completed = true
-			this _condition broadcast()
-		)
+		this _mutex lock()
+		this _completed = true
+		this _condition broadcast()
+		this _mutex unlock()
```
Comment (sebastianbaginski) [inline source/concurrent/Promise.ooc]: This one puzzles me. When using `with()`, one of the tests in other repo reports more race conditions.
But the main reason I changed - I think it's not worth creating / deleting a temporary closure object with `calloc / free` just to save few characters (not even a line).

File: source/concurrent/ThreadPool.ooc (hunk 1 above)
Comment (emilwestergren) [inline source/concurrent/ThreadPool.ooc]: Here is a good case to use `this _mutex with(|| status = this _state == _PromiseState Finished)` though, but doesn't matter to me.

Comment (sebastianbaginski) [inline source/concurrent/ThreadPool.ooc]: Hmm that's true, it's much cleaner. @marcusnaslund ?

Comment (marcusnaslund) [inline source/concurrent/ThreadPool.ooc]: Agree with @emilwestergren

Comment (marcusnaslund) [inline source/concurrent/ThreadPool.ooc]: unless there's a cost of doing it

Comment (emilwestergren) [inline source/concurrent/ThreadPool.ooc]: There is, a closure needs to allocated and freed.

Comment (sebastianbaginski) [inline source/concurrent/ThreadPool.ooc]: `calloc / free` on a closure

**Pull request merged**